### PR TITLE
BUG: replace with regex raising for StringDType

### LINKED ISSUE #35977: BUG: replacement works for object but not string dtype
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample

```python
import pandas as pd
import numpy as np

a = pd.DataFrame({'a': ['a', 'b', 'c'], 'b': ['d', '', '']}, dtype='object')
b = pd.DataFrame({'a': ['a', 'b', 'c'], 'b': ['d', '', '']}, dtype='string')

print(a)
a.replace(r'^\s*$', pd.NA, regex=True, inplace=True)
print(a)

print(b)
b.replace(r'^\s*$', pd.NA, regex=True, inplace=True)
print(b)
```

#### Problem description

replace(r'^\s*$', pd.NA, regex=True, inplace=True) works on object dtype, but not on stringdtype

#### Expected Output

Same on both replacement

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here leaving a blank line after the details tag]

</details>

### LINKED ISSUE #41333: BUG: Unexpected behavior pandas.DataFrame.replace with "string" dtype
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---
I'm trying to remove certain characters in a DataFrame, in this case, I am trying to remove the **ñ**.
#### Code Sample

```python
df = pd.DataFrame([['añoos','asññ','ÑolÑÑss'],['añoos','asññ','ÑolÑÑss']], dtype='string')
df.replace({'ñ':'n',
            'Ñ':'N'}, 
           regex = True, inplace = True)
```
The output is:

|    | 0     | 1    | 2       |
|---:|:------|:-----|:--------|
|  0 | añoos | asññ | ÑolÑÑss |
|  1 | añoos | asññ | ÑolÑÑss |

#### Problem description

If I do not specify the dtype to ```string```, the output is the expected as follows:

#### Expected Output

```python
df = pd.DataFrame([['añoos','asññ','ÑolÑÑss'],['añoos','asññ','ÑolÑÑss']])
df.replace({'ñ':'n',
            'Ñ':'N'}, 
           regex = True, inplace = True)
print(df)
```

|    | 0     | 1    | 2       |
|---:|:------|:-----|:--------|
|  0 | anoos | asnn | NolNNss |
|  1 | anoos | asnn | NolNNss |
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.8.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : es_419.UTF-8
LOCALE           : Spanish_Colombia.1252

pandas           : 1.2.4
numpy            : 1.19.4
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 54.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : 0.9.3
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 2.11.3
IPython          : 7.22.0
pandas_datareader: None
bs4              : 4.6.0
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : None
odfpy            : None
openpyxl         : 3.0.6
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.52.0

</details>

### PR DESCRIPTION
**Pull request by mzeitlin11**
- [x] closes #35977, closes #41333
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

2 separate issues at play here - regex was ignored in #41333 (specific to the `replace_list` path) but then if `replace_regex` was hit with anything but an object type that would just raise.

### CODE REVIEW

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -744,7 +744,7 @@ Strings
 ^^^^^^^
 
 - Bug in the conversion from ``pyarrow.ChunkedArray`` to :class:`~arrays.StringArray` when the original had zero chunks (:issue:`41040`)
--
+- Bug in :meth:`Series.replace` and :meth:`DataFrame.replace` either raising or ignoring replacements with ``regex=True`` for ``StringDType`` data (:issue:`41333`, :issue:`35977`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.3.0.rst:747]: it's only raising on master. in 1.2.4 the replace silently failed. the release note need only describe the change from the released behavior.

```suggestion
- Bug in :meth:`Series.replace` and :meth:`DataFrame.replace` ignoring replacements with ``regex=True`` for ``StringDType`` data (:issue:`41333`, :issue:`35977`)
```

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @mzeitlin11 generally lgtm for the fix in hand, although we should probably be dispatching to the `_str_replace` method on the EA.

**Comment by mzeitlin11** [comment]
> Thanks @mzeitlin11 generally lgtm for the fix in hand, although we should probably be dispatching to the `_str_replace` method on the EA.

Agreed, but I think that decision could be handled independently of fixing this. There's some additional complexity because I think `_str_replace` assumes replacement occurs with a string, which is not guaranteed by the general replace function (though it would be a strange thing to do).

**Comment by simonjayhawkins** [comment]
> Agreed, but I think that decision could be handled independently of fixing this.

sure.

**Review by simonjayhawkins — APPROVED** [review]
lgtm pending green

**Comment by jreback** [comment]
thanks @mzeitlin11 very nice!

**Pull request merged**